flamedisx/lxe_blocks/reconstruct_signals.py

```diff
@@ -0,0 +1,252 @@
+import typing as ty
+
+import numpy as np
+from scipy import stats
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+import flamedisx as fd
+export, __all__ = fd.exporter()
+o = tf.newaxis
+
+
+class ReconstructSignals(fd.Block):
+    """Common code for ReconstructS1 and ReconstructS2"""
+
+    model_attributes = ('check_acceptances',)
+    non_integer_dimensions = ('s1_raw', 's2_raw',)
+
+    # Whether to check acceptances are positive at the observed events.
+    # This is recommended, but you'll have to turn it off if your
+    # likelihood includes regions where only anomalous sources make events.
+    check_acceptances = True
+
+    # Prevent pycharm warnings:
+    source: fd.Source
+    gimme: ty.Callable
+    gimme_numpy: ty.Callable
+
+    raw_signal_name: str
+    signal_name: str
+
+    def _simulate(self, d):
+        bias = self.gimme_numpy(f'reconstruction_bias_{self.signal_name}_simulate',
+                                bonus_arg=d[self.raw_signal_name].values)
+        mu = d[self.raw_signal_name] * bias
+
+        # clipping this to (1e-15, float32max) to be symmetric with _compute
+        smear = self.gimme_numpy(f'reconstruction_smear_{self.signal_name}_simulate',
+                                 bonus_arg=d[self.raw_signal_name].values)
+        smear = np.clip(smear, 1e-15, tf.float32.max)
+        # TODO: why some raw signals <=0?
+        # checked 1e7 events and didn't see any raw_signals<=0..
+
+        d[self.signal_name] = stats.norm.rvs(
+            loc=mu,
+            scale=smear)
+        # Call add_extra_columns now, since s1 and s2 are known and derived
+        # observables from it (cs1, cs2) might be used in the acceptance.
+        # TODO: This is a bit of a kludge
+        self.source.add_extra_columns(d)
+        d['p_accepted'] *= self.gimme_numpy(self.signal_name + '_acceptance')
+
+    def _annotate(self, d):
+        # should really be rading reconstruction bias and smear at the raw
```

Comment:
I think we discussed this - did you not say that there's a map you can read from to get the bias/smear as a function of reconstructed value also? Though I'm now a bit confused by this, because the map from raw -> final is probabilistic, not deterministic... Could you clarify?

Reply:
This just looks like an unfortunate comment. The `annotate` only has to estimate bounds, it is OK that the actual mapping is probabilistic. 

Nonetheless, the code provides the possibility to use different model functions for the raw->recon mapping (used in simulate and compute) and the recon->raw mapping (used in annotate).